token-metadata/program/src/processor/burn/nonfungible.rs

```diff
@@ -0,0 +1,130 @@
+use super::*;
+
+pub(crate) struct BurnNonFungibleArgs {
+    pub(crate) metadata: Metadata,
+}
+
+pub(crate) fn burn_nonfungible(ctx: &Context<Burn>, args: BurnNonFungibleArgs) -> ProgramResult {
+    let edition_info = ctx.accounts.edition_info.unwrap();
+
+    // If you're passing in parent accounts for a nonfungible you're using this handler wrong.
+    // Parent accounts are only for burning nonfungible editions.
+    if ctx.accounts.parent_mint_info.is_some()
+        || ctx.accounts.parent_edition_info.is_some()
+        || ctx.accounts.parent_token_info.is_some()
+        || ctx.accounts.edition_marker_info.is_some()
+    {
+        return Err(MetadataError::InvalidParentAccounts.into());
+    }
+
+    // If the NFT is a verified part of a collection but the user has not provided the collection
+    // metadata account, we cannot burn it because we need to check if we need to decrement the collection size.
+    if ctx.accounts.collection_metadata_info.is_none()
+        && args.metadata.collection.is_some()
+        && args.metadata.collection.as_ref().unwrap().verified
+    {
+        return Err(MetadataError::MissingCollectionMetadata.into());
+    }
+
+    let edition_account_data = edition_info.try_borrow_data()?;
+
+    // First byte is the object key.
+    let key = edition_account_data
+        .first()
+        .ok_or(MetadataError::InvalidMasterEdition)?;
+    if *key != Key::MasterEditionV1 as u8 && *key != Key::MasterEditionV2 as u8 {
+        return Err(MetadataError::NotAMasterEdition.into());
+    }
+
+    // Next eight bytes are the supply, which must be converted to a u64.
+    let supply_bytes = array_ref![edition_account_data, 1, 8];
+    let supply = u64::from_le_bytes(*supply_bytes);
```

Comment:
I assume this manual deserialization is so that you don't have to separately try to deserailize V1 or V2?

Reply:
That's one benefit, though we're probably safe just deserializing V2s. Also, it's cheaper than deserializing the whole struct, but I don't have hard data on how much cheaper. 

It's only 136 bytes more code to deserialize piecemeal this way, so seems ok to do for now.

Reply:
Sounds good to me